spring-cloud-stream-binder-rabbit-core/src/main/java/org/springframework/cloud/stream/binder/rabbit/provisioning/RabbitExchangeQueueProvisioner.java

```diff
@@ -0,0 +1,510 @@
+ * limitations under the License.
+ */
+
+package org.springframework.cloud.stream.binder.rabbit.provisioning;
+
+import java.util.HashMap;
+import java.util.Map;
+
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+
+import org.springframework.amqp.AmqpConnectException;
+import org.springframework.amqp.core.AnonymousQueue;
+import org.springframework.amqp.core.Binding;
+import org.springframework.amqp.core.BindingBuilder;
+import org.springframework.amqp.core.DirectExchange;
+import org.springframework.amqp.core.Exchange;
+import org.springframework.amqp.core.ExchangeBuilder;
+import org.springframework.amqp.core.FanoutExchange;
+import org.springframework.amqp.core.Queue;
+import org.springframework.amqp.core.TopicExchange;
+import org.springframework.amqp.rabbit.connection.ConnectionFactory;
+import org.springframework.amqp.rabbit.core.RabbitAdmin;
+import org.springframework.beans.factory.config.ConfigurableListableBeanFactory;
+import org.springframework.beans.factory.support.DefaultListableBeanFactory;
+import org.springframework.cloud.stream.binder.ExtendedConsumerProperties;
+import org.springframework.cloud.stream.binder.ExtendedProducerProperties;
+import org.springframework.cloud.stream.binder.rabbit.properties.RabbitCommonProperties;
+import org.springframework.cloud.stream.binder.rabbit.properties.RabbitConsumerProperties;
+import org.springframework.cloud.stream.binder.rabbit.properties.RabbitProducerProperties;
+import org.springframework.cloud.stream.provisioning.ConsumerDestination;
+import org.springframework.cloud.stream.provisioning.ProducerDestination;
+import org.springframework.cloud.stream.provisioning.ProvisioningProvider;
+import org.springframework.context.support.GenericApplicationContext;
+import org.springframework.util.StringUtils;
+
+/**
+ * AMQP implementation for {@link ProvisioningProvider}
+ *
+ * @author Soby Chacko
+ */
+public class RabbitExchangeQueueProvisioner implements ProvisioningProvider<ExtendedConsumerProperties<RabbitConsumerProperties>,
+		ExtendedProducerProperties<RabbitProducerProperties>> {
+
+	private static final AnonymousQueue.Base64UrlNamingStrategy ANONYMOUS_GROUP_NAME_GENERATOR
+			= new AnonymousQueue.Base64UrlNamingStrategy("anonymous.");
+
+	/**
+	 * The delimiter between a group and index when constructing a binder
+	 * consumer/producer.
+	 */
+	private static final String GROUP_INDEX_DELIMITER = ".";
+
+	protected final Log logger = LogFactory.getLog(getClass());
+
+	private final RabbitAdmin rabbitAdmin;
+
+	private final GenericApplicationContext autoDeclareContext = new GenericApplicationContext();
+
+	public RabbitExchangeQueueProvisioner(ConnectionFactory connectionFactory) {
+		this.rabbitAdmin = new RabbitAdmin(connectionFactory);
+		this.autoDeclareContext.refresh();
+		this.rabbitAdmin.setApplicationContext(this.autoDeclareContext);
+		this.rabbitAdmin.setIgnoreDeclarationExceptions(true);
+		this.rabbitAdmin.afterPropertiesSet();
```

Comment:
IMO we address this as a separate concern through another issue. This PR already moved a bunch of code around. It might be better to merge it and then address this later.